### PR TITLE
Make TrimmerDefaultAction public

### LINKED ISSUE #16140: Make _TrimmerDefaultAction public
In .NET 6, we are introducing `[assembly: AssemblyMetadata("IsTrimmable", "True")]` as a way to opt a library into trimming. MSBuild metadata can also be used to override this per-assembly.

With this change, the linker has two global knobs to control trimming:
- `--trim-mode`: the trimming granularity for assemblies marked `IsTrimmable`
  - The SDK sets this to `copyused` by default, but it is configurable via the property `TrimMode`
- `--action`: the trimming granularity for assemblies *not* marked `IsTrimmable`
  - The SDK sets this to `copy` by default (so that assemblies which aren't marked `IsTrimmable` are kept).

We should expose a public MSBuild option that maps to the global `--action` (maybe `TrimmerDefaultAction`), which allows developers to enable member-level trimming for the entire application by setting:
```xml
<TrimMode>link</TrimMode>
<TrimmerDefaultAction>link</TrimmerDefaultAction>
```

When we discussed full-app trimming earlier, we were hesitant to make it *too* easy (see the notes under [`TrimAllAssemblies`](https://github.com/mono/linker/blob/main/docs/design/trimmed-assemblies.md#trimallassemblies-global-opt-in)).

On the other hand, the option is already there (introduced in https://github.com/dotnet/sdk/pull/16094), so the real question is just whether it should be exposed to the user (and if so, what the name should be). Note that xamarin-android already has some user-facing knobs to do exactly this (`<AndroidLinkMode>Full</AndroidLinkMode>`).

@marek-safar @vitek-karas @samsp-msft @eerhardt

### PR DESCRIPTION
**Pull request by sbomer**
Fixes https://github.com/dotnet/sdk/issues/16140
Relevant for https://github.com/dotnet/docs/pull/23766

### CODE REVIEW

**Comment by eerhardt** [comment]
Are we ready to expose this publicly? We now have a really easy way to "full trim" your app. Are we ready to support this mode?

**Comment by sbomer** [comment]
That would be my recommendation, to get the behavior described in the WIP library trimming [docs](https://github.com/dotnet/docs/pull/23766):

> `<TrimmerDefaultAction>link</TrimmerDefaultAction>` ensures that only used parts of dependencies are analyzed. Without this option, you would see warnings originating from _any_ part of a dependency that doesn't set `[AssemblyMetadata("IsTrimmable", "True")]`, including parts that are unused by your library.

I can note that this isn't recommended for app authors yet (or just not mention it anywhere else in the docs). If we really don't want to expose this I can change the library mode recommendation to:

```xml
<Target Name="TrimDependencies" BeforeTargets="PrepareForILLink">
  <ItemGroup>
    <ManagedAssemblyToLink IsTrimmable="true" />
  </ItemGroup>
</Target>
```

**Comment by vitek-karas** [comment]
While I do agree that we don't want people to use this too often (as it comes with a relatively high risk), I think that the current way to do this (via target) looks like a big hack.

Maybe we should change it so that when this is set, we always enable trim analysis warnings (even if SDK says otherwise). This would also get us the desired behavior where if I'm in Xamarin and I ask for full trimming, I get warnings by default (it being an advanced scenario).

**Comment by sbomer** [comment]
> when this is set, we always enable trim analysis warnings (even if SDK says otherwise).

This sounds like a good default, but I think it should be up to each SDK to decide (instead of the .NET SDK overriding what the Xamarin SDK or user tells us). Xamarin users who have been successful with LinkAll might not be interested in the warnings.

I think a good way to accomplish this would be for the Xamarin SDKs to suppress warnings only when `TrimmerDefaultAction` is `copy`. If we agree, I can file issues recommending this.

**Comment by marek-safar** [comment]
> Xamarin users who have been successful with LinkAll might not be interested in the warnings.

It's quite unlikely that app that works with LinkAll today (presumably because it also has several things manually persisted) will just work after migration to net6.

I think we should enable the warnings, especially if we issue only 1 per assembly by default, for anyone trimming with "link-all" setup.

**Review by marek-safar — APPROVED** [review]

**Comment by eerhardt** [comment]
> Maybe we should change it so that when this is set, we always enable trim analysis warnings (even if SDK says otherwise). This would also get us the desired behavior where if I'm in Xamarin and I ask for full trimming, I get warnings by default (it being an advanced scenario).

I like this idea. 👍.

**Comment by sbomer** [comment]
Should there be a way to override that default (a way to to turn off warnings when using LinkAll)?

**Comment by vitek-karas** [comment]
I would expect the user provided value for `SuppressTrimAnalysisWarnings` to always be honored. But I guess this is pretty tricky with SDK interactions... can we make that work anyway?

**Comment by sbomer** [comment]
It might be possible by setting a default in early properties, then detecting if it has been changed after SDK props are imported... but I am not a fan of that pattern. I think just updating the places that set SuppressTrimAnalysisWarnings to true (so they only do it when the TrimmerDefaultAction is copy) is a simpler way to get the behavior we want.

**Comment by vitek-karas** [comment]
Alternatively we could change the SDKs to set a private property instead and combine the public property with the private property in one place, taking the TrimmerDefaultAction into consideration.

**Comment by marek-safar** [comment]
> a way to to turn off warnings when using LinkAll

Why would not standard warning (1 in this case) suppression work?

**Comment by vitek-karas** [comment]
Which "1" warning so you have in mind. The idea is that is I turn on agressive trimming for everything by default I get the detailed warnings - that might be 100s for complex apps. The end of the discussion is if there's a way to suppress them - they are suppresable via NoWarn, but the list is like 80 different warning codes, so not ideal.

**Comment by sbomer** [comment]
That would work for nuget dependencies of the app, but we still show detailed warnings for app code (with different codes).

**Comment by sbomer** [comment]
> The idea is that is I turn on agressive trimming for everything by default I get the detailed warnings

Just to clarify - we would still collapse warnings from nugets, right?

**Comment by vitek-karas** [comment]
yes - sorry I didn't make that clear

**Comment by marek-safar** [comment]
> The idea is that is I turn on agressive trimming for everything by default I get the detailed warnings - that might be 100s for complex apps. The end of the discussion is if there's a way to suppress them - they are suppresable via NoWarn, but the list is like 80 different warning codes, so not ideal.

I think the scenario is a bit different. You turn it on and you have code you can modify (basically you have multiple ways to resolve the warnings) and you have tons of dependencies you cannot modify. If you are going to publish your app with aggressive trimming all unaddressed warnings are "wrong" and you would like to suppress them. I'd suspect these will come from the nuget assemblies and my understanding is that we are going to fold the trim analysis warning from them by default. Any warnings in the app code can be suppressed with pragma(s) or resolved.

**Comment by sbomer** [comment]
> I'd suspect these will come from the nuget assemblies and my understanding is that we are going to fold the trim analysis warning from them by default. Any warnings in the app code can be suppressed

This is still true with the latest change, and will be the recommended way to suppress warnings (though note that pragmas don't work for trim warnings) - so I think we agree on this.

The question is only about `SuppressTrimAnalysisWarnings`, which suppress warnings even from the app, and was set by default for some SDKs in https://github.com/dotnet/sdk/pull/16327. The latest [commit](https://github.com/dotnet/sdk/pull/16865/commits/df910817f065ea7fa095ad23a4185fbcb3fcb409) just avoids suppressing them for "LinkAll". I think this matches what you were asking for in https://github.com/dotnet/sdk/pull/16865#issuecomment-820574681:

> I think we should enable the warnings, especially if we issue only 1 per assembly by default, for anyone trimming with "link-all" setup.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets [hunk 1]
```diff
@@ -48,8 +48,8 @@ Copyright (c) .NET Foundation. All rights reserved.
   <PropertyGroup Condition="'$(SuppressTrimAnalysisWarnings)' == '' And ('$(EnableTrimAnalyzer)' == 'true' Or '$(PublishTrimmed)' == 'true')">
     <!-- Trim analysis warnings are suppressed for .NET < 6. -->
     <SuppressTrimAnalysisWarnings Condition="$([MSBuild]::VersionLessThan('$(TargetFrameworkVersion)', '6.0'))">true</SuppressTrimAnalysisWarnings>
```
Comment (vitek-karas) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:50]: Should allow the user to override this even in <6? Currently this ignores the existing value of the property...

**Review by vitek-karas — APPROVED** [review]

File: src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.props [hunk 2]
```diff
@@ -30,7 +30,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     <PublishTrimmed Condition="'$(PublishTrimmed)' == ''">true</PublishTrimmed>
     <TrimMode Condition="'$(TrimMode)' == ''">link</TrimMode>
     <TrimmerRemoveSymbols Condition="'$(TrimmerRemoveSymbols)' == ''">false</TrimmerRemoveSymbols>
-    <SuppressTrimAnalysisWarnings Condition="'$(SuppressTrimAnalysisWarnings)' == ''">true</SuppressTrimAnalysisWarnings>
+    <SuppressTrimAnalysisWarnings Condition="'$(SuppressTrimAnalysisWarnings)' == '' And '$(TrimmerDefaultAction)' != 'link'">true</SuppressTrimAnalysisWarnings>
```
Comment (eerhardt) [inline src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.props:33]: Note that these `.props` files get evalutated before the .csproj is evaluated.

So if I have a Blazor or Web .csproj that contains:

```xml
<TrimmerDefaultAction>true</TrimmerDefaultAction>
```

This check is going to run too soon, and the warnings will still be suppressed.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets (hunk 1 above)
Comment (sbomer) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:50]: It is possible to override in .NET 5 - the whole propertygroup is skipped if the user already set it.

File: src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.props (hunk 2 above)
Comment (sbomer) [reply src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.props]: Thanks for catching this, I moved it to the targets.

**Pull request merged**